proposed/2021/VulnerabilityInfoInPMUI.md

```diff
@@ -0,0 +1,52 @@
+* Make any improvements to the package deprecation experience.
+* Change the registration API to pull in more details about a package vulnerability. This will be considered in a future iteration.
+* Show a warning icon in the Solution Explorer when packages with vulnerabilities are installed. This will be considered in a future iteration.
+
+## Solution
+
+A list of vulnerabilities for a package is read from the server, along with other data, whenever a package version is selected (DetailControlModel.SetCurrentPackageAsync) or when the selection version changes (DetailControlModel.SelectedVersionChangedAsync). This data is stored in the DetailedPackageMetadata and is used by the Details Pane. The vulnerabilities information exposed in the PackageSolutionDetailControlModel which implements DetailedPackageMetadata is currently limited to whether a vulnerability exists and what the maximum severity of the vulnerability is. This will be changed to expose instead a list of vulnerabilities, their severity, and their advisory links.
+
+The Details Pane will have a new section in the view called "Vulnerabilities" that will be visible when a package has one or more vulnerabilities. The Vulnerabilities section will display the highest severity of the vulnerabilities and show a list of vulnerabilities, their severity, and a link to the advisory. To stay consistent with the current design of the details pane, the vulnerabilities section will be created as its own UserControl that will be used by the DetailControl.
+
+The PackageManagerTopPanel will be modified to handle showing a warning icon when there are either deprecated or installed packages, or both. A new method will be added to this class to handle getting the count of the vulnerable packages installed which will be used in the tooltip and also to determine whether or not to show the warning icon for vulnerabilities.
+
+The PackageItemViewModel will have a new method to query the Vulnerability status of a given package asynchronously, much like it currently reads package deprecation status. This information will be used, along with the deprecation status, to determine if the warning icon should appear by the package in the package list. The tooltip for the icon will be updated to display information about the vulnerabilities and/or deprecated packages.
+
+The NuGetPackageSearchService will be updated to query the Vulnerability information for a given package so it can be used to by the PackageItemViewModel to know about vulnerabilities. Some code may need to be refactored here so the deprecation data and the vulnerability data is pulled in by the same query as package metadata and then separated into deprecation and vulnerability data by the caller rather than duplicating the call to the service for each chunk of data.
+
```

Comment:
I don't know if deprecation already handles this, but when the sources list is set to "all", this is a special case in PM UI, where when multiple sources contain the same package ID, we join them into a single entry. NuGet is built on the assumption that all packages are immutable and globally unique, so when this is violated and package description or package icon is different on different sources, it's effectively random which package metadata you'll actually get.

I really hope that package deprecation doesn't do this, but if it doesn't, I think it's really important for vulnerabilities to ensure that they're all listed, since security is so important.

In the case where multiple package sources list vulnerabilities, how will the list be merged?  If two sources list the same CVE/AdvisotyUrl, will the CVE be listed once or twice? If they list the same CVE, but for some reason have different severities, how will this be handled?

Reply:
@chgill-MSFT is vulnerability information available in any feed other than nuget.org?

Reply:
> is vulnerability information available in any feed other than nuget.org?

I don't think that's the point. The point is if 2 sources have *the same package*, we need to ensure the one with vulnerability flags (even if it's only nuget.org), is prioritized/loaded and not the one that doesn't contain vulnerability data at all.

Reply:
@jebriede can you check the code to see how/if this is handled for deprecation?

@joelverhagen do/can other feeds have their own vulnerability information? My guess is no since I believe the GitHub data is NuGet.org-specific.

@zivkan my understanding is that the PMUI will show the package details from whichever feed responds first. Is that correct? If so, do we continue to listening for a response from the other feeds after one has responded?

I'm wondering if we can show the vulnerability/deprecation information from NuGet.org even if it's not the first feed to respond.

Reply:
> do/can other feeds have their own vulnerability information? My guess is no since I believe the GitHub data is NuGet.org-specific.

It is technically possible. NuGet.org simply fills out some documented properties in JSON API responses with vulnerability information. Any V3 feed could do this, sourcing info however they choose. AFAIK there is no logic in NuGet client to limit vulnerability information (or deprecation, or prefix reservation, etc)  from coming from NuGet.org -- which is good.

Reply:
> my understanding is that the PMUI will show the package details from whichever feed responds first. Is that correct? If so, do we continue to listening for a response from the other feeds after one has responded?

I don't know. Someone who knows more can reply, otherwise someone working on the feature needs to investigate. I'm just saying I think this is a scenario that this spec should address.

> I'm wondering if we can show the vulnerability/deprecation information from NuGet.org even if it's not the first feed to respond.

With enough code changes, even if NuGet today is ignoring responses after the fastest server responds first, we can change it. Might be costly if the software design doesn't make this easy, in which case it depends on valuable it is to the PMs & division.

Reply:
For the sake of the MVP I think it is out of scope to try and resolve an existing issue with how the "All" package sources code decides to take metadata from only one package source and ignore the rest. If vulnerabilities does not work when the All package source is selected in the PMUI, we can make note of that in our release notes and fix that in the future. I'd like to get an opinion from @aortiz-msft on the matter.

Reply:
I've replied to your offline thread. 
I don't believe it's an edge case or out of scope to consider whether multiple sources could present a scenario where the "win" misses vulnerability data. An MVP for vulnerabilities should present consistent/reliable warnings to the user. Otherwise to me, it really takes the "viable" out of MVP.

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs#L206

Reply:
Eyeballing the code, it seems possible that these secondary queries are all-or-none, so there may not be a problem. Something you can (should, IMO) do is write a test that shows how, given multiple sources with and without vulnerability data, the UI is always saturated with the vulnerability data.

Reply:
@donnie-msft thanks for pointing out the code that deprecation uses to handle merging the deprecation status by always ensuring the non-null data is available. I misunderstood some of Andy's comment to mean that we don't have a mechanism for merging the contents but I see now that's not what he was saying.

In the case where there are two sources that both provide vulnerability data (so they're both non-null), we'll want to take the one that exposes the highest severity of vulnerability and not just take the first vulnerability metadata that is not null like deprecation is doing. That would at least expose the vulnerabilities with the highest severity.

I'll add a line to the spec about merging vulnerability data for the All sources case.